### DISCUSSION TITLE
Wiring rework fixing inconsistency and unintuitiveness

### DISCUSSION THREAD

**Discussion by Working-Joe** [FakeFishGames/Barotrauma#5001] [Ideas and suggestions]
Wiring rework fixing inconsistency and unintuitiveness (Essentially just for my own future reference)

_TL:DR just check the headings and the bullet points_

_I'd love to hear feedback on whether I've overlooked anything that might break with this system, as well as ideas on how to improve it. :)_

### Problems with the current wiring system
The current wiring system has a number of inconsistencies which make it rather hard for players to learn how to wire. Most of these problems stem from differences in which the components are coded. 

To name a few problems:
- The update order of components causes delays in signal propagation, which can lead to wrong values or incorrect signal flow. (See also the readme [on this repo](https://github.com/Jlobblet/Barotrauma-Circuit-Resolver))
- Some components output constantly on their update, others output as soon as they receive a signal.
- Some components reset their inputs when receiving no signal, others remember the last input
- Time frames have different meanings on different components. Some components send their false output once they time out, others send nothing.
- Some components (which send on their update) can receive multiple inputs in a single frame, but only process the last received ones before their update. 
- Any arithmetic component for some reason (bug?) starts with their inputs true for the duration of their timeframe. This can not be patched easily because it currently provides the functionality of detecting a round start. 

These inconsistensies/hidden component behaviours can be confusing to new players, but can also make circuits needlessly complex as you constantly need to account for possible delays in the signal propagation. 

### The ideal solution
Ideally, the system would be changed so that 
- Time frames are removed, as they seemingly exist essentially as a fix relating to the update order. 
- All components will send an output if updated *or* All components will send an output upon receiving input
- The component system is overhauled so that most items send once they receive a signal and not when they are updated

This would clear up any if not all inconsistencies and make the component system simpler and more intuitive to use, without sacrificing capabilities. 

### Rework proposal
I've considered a number of different reworks, but I think this one is the best combination of efficiency and intuitiveness:

- components send once they have both inputs, then remain inactive/blocked for the rest of the frame.
- Allow one wire per input pin on components.
- Timeframes are removed, instead components remember their inputs indefinitely. 
- null signals will be sent to close the next component (instead of sending no signal)
- Unconnected inputs are considered received (either 0 or the value filled in the property field if e.g. `set_output`)
- Add merge/combine component that has multiple inputs, and once received sends one of them through per frame and let one through depending on a priority order. For instance, if signal_2 and signal_3 are received, only signal_2 is sent through. 
- Add buffer component that holds on to signal for a limited time and then returns to sending nothing/a false output to provide timeframe functionality.
- Add round start component that sends a signal whenever a sub is loaded on round start (essentially the functionality arithmetic components now provide with their bugged inputs).

During each frame, the components are updated as follows:

- All source components (everything that outputs a state: mems, delays, relays (state_out), buffers, osc, inputs) are updated one by one. 
- Upon update, the source components send the signal they received last frame.
- source components store their received values during a frame but do not update their output until the next frame.

### motivation

Existing components are already designed to process no more than one received set of inputs per frame. Multiple connections are used reliably if and only if one of the components will send a signal into the input during a frame. Any further input received before the `Update()` method are overwritten.
Allowing only one input ensures this overwritingdoes not happen. Signals can still be combined using the new priority merge/combine component. Furthermore, since the components send as soon as they receive both inputs, the system will be update order independent and there will be no propagation delays or need for a timeframe. 
Any functionality brought by timeframes aside from fixing the update order problems will be brought by the other two new components.

Furthermore, since memory components effectively update their value at the same time (by taking on the value of the previous frame), it would even be possible to make reliable shift registers by chaining memory components to work as a 1 frame delay.  

### discussion

This system does come with two trade-offs. The first downside of this system is that loops not containing a source component will not work properly. However, these loops are with few exceptions unstable (the signal would loop forever) and do not work correctly in the current wiring implementation either. 

A second trade-off relates to performance This decrease is because with these changes each component will send an output regardless of wether they receive an input - in other words, each wire will have a signal flowing over it on each frame. This is necessary to resolve the circuitry at nodes where signals are combined (the new merge/combine component)

However, I believe this decrease is marginal for a few reasons:
- The majority of components already calculate and send an output each frame
- "inactive" components that receive two null signals will have minimal load since they can immediately send their last output through
- the amount of signals on a wire in a frame is limited (whereas it is currently not limited)

### implementation
Lastly, an update with these changes can not be pushed through without reverse compatibility issues. One option could be to add the changes as entirely new components, while the old components continue their existence as legacy components. 

A more elegant solution may be combining a change like this with an idea like circuit boxes. (see also #4721). The new components could then only be placed within the circuit boxes without interfering with existing legacy components.

This would have further benefits for performance as well. It might be possible to put circuit boxes "to sleep" if they receive no input or depending on the activity inside the box. If none of the sources or inputs changes value, the box will not have to do calculations. It could even be possible to give the boxes a changeable update frequency with an `update` input you could hook an oscillator to. And if you really have a dev that needs to be punished, it could technically be possible to compile the circuit boxes and optimise the code. 

But there are even more advantages that could be gained from these boxes. To name a few:
- Circuits can be integrated in subs much, much more elegantly. Currently the most common solutions are to just hide them in walls or outside the sub with "Hidden in game" ticked. This means that the circuits can not be accessed. 
- The components could be removed from the shop and fabrication UI to reduce clutter.
- There would be a great reduction in wire rendering, which is a considerable part of rendering load.
- The circuit box interface could be locked away from griefers in one place
- Circuits could be balanced by limiting the amount of components in one circuit box

### Circuit boxes
To add to the idea of adding these components as separate components inside a circuit box, I made a few mock-up images I'm putting here to show what such a solution could look like.
![image](https://user-images.githubusercontent.com/50582661/122674250-8d5ebc00-d1d4-11eb-88f7-328b69f7a0bb.png)

The circuit box will essentially be a subgrid of components that lives within the box. In here you can wire things up as if in wiring mode within the editor. Components and wires can be placed at no cost, up to a certain component limit. There could be boxes of different sizes which could hold more components. 
For the inputs and outputs there would be separate "IO" components, which relay an input or output of the circuit box similar to how WiFi components relay outputs. In the mock-up these the blue numbered components are the IO components.

![image](https://user-images.githubusercontent.com/50582661/122674534-e713b600-d1d5-11eb-9756-de8117a5d258.png)

Editing the component values would go as usual, with the normal in game (or in-editor) info box for the wiring showing up. 

![image](https://user-images.githubusercontent.com/50582661/122674607-2e01ab80-d1d6-11eb-9b30-2868cc379c8b.png)

Because the components are now cost-free, they can be placed using a search bar similar to in the editor - in fact, I reckon a lot of the UI code can just be stolen from the sub editor code. \:)

![image](https://user-images.githubusercontent.com/50582661/122674559-fabf1c80-d1d5-11eb-8949-113703750790.png)

Because the wiring is now integrated in the UI, it would be possible to add a lot of tooltips to the interface for accessibility. Things like showing how to delete a connection, or highlighting the hotkeys.

### Notes

- The current existance of logical components sending no output could also be considered unintuitive, as it is different from the way gates behave in real life. However, I've chosen to keep this behaviour as it adds a lot of functionality as part of allowing custom outputs. People can always sets their outputs and false outputs to 0 and 1 respectively if they want the original electrical engineering experience™
- Inputs of the circuit box also can have at most 1 input pin. It might be sensible to add a component to allow aggregating similar signals at the input, for instance if you're checking if any component outputs 1 on a series of detectors. This to prevent the need to make large 
- WiFi components can receive multiple inputs, which would break behaviour inside circuit boxes. It seems sensible to allow these to be placed only outside boxes, as they're technically part of I/O. It also doesn't make logical sense that a wifi box communicates with itself internally through WiFi.
- Because Relays are made to relay power and there is no power in the circuit boxes, it would be logical to make a separate SwitchComponent inside the boxes that only has the signal inputs. This component would also have inputs that accept only one wire. To separate the functionality of signal distribution trees it may also be nice to make a separate DistributeComponent, rather than having relays that are always on for this.

**Comment by mLuby** [comment]
Interesting proposal, @Working-Joe. Here are some thoughts:

1. > Components send once they have both inputs…
    > Allow one input per component.
    
    I'm having trouble visualizing how this would work for components like AND or Relay. Wouldn't there need to be special cases for the merge components?

1. > Add round start component that sends a signal whenever a sub is loaded on round start (essentially the functionality arithmetic components now provide with their bugged inputs).

    This feels like a good use for the existing Memory component with preset values, instead of adding a new component.
    
1. > Any further input [signals] received before the `Update()` are overwritten. Allowing only one input ensures this overwriting does not happen. Signals can still be combined using the new priority merge/combine component.

    Are you talking about a case like this, with multiple continuous signals (Mem=3 and Mem=5) tied to the same input pin (in2) of a single component (Add)?
    ```
    Mem=1.out->in1.Add
    Mem=3.out->in2.Add.out -> 4 or 6 or 9???
    Mem=5.out->in2.Add
    ```
    If that's right, I think it's entirely reasonable that this behavior is undefined/buggy. Maybe I'm misunderstanding the problem you're trying to solve.
    
1. > Each component will send an output regardless of whether they receive an input—in other words, each wire will have a signal flowing over it on each frame.

    They say "don't let perfect be the enemy of the good;" still, I think this point deserves further consideration. Ideally only _changes_ in signals need to propagate. Wouldn't that be possible if every component remembered its inputs, and only forgot them on receiving one or more `null` signals? It would mean that interactive components like doors and hatches may need to become uninteractable on receiving a non-null signal.

1. > The ideal solution…

    I agree, and I'd add a fourth: drop the notion of continous signals entirely. When I experienced the learning curve, and from what I've noticed from helping others, is that the distinction between pulsed and continuous signals is confusing. I think if everything were one way or the other it would be simpler to reason about circuit behavior, and for reasons above, pulsed is the more performant option.

1. > One option could be to add the changes as entirely new components, while the old components continue their existence as legacy components.

    You're right that this is a breaking/non-backward-compatible change no matter what, but I'd be quite wary of keeping _multiple distinct_ wiring/update/component systems in the game at the same time. And I think the breaking changes could be relatively few and obscure (though [Hyrum's Law](https://www.hyrumslaw.com) has already been demonstrated by this community).

**Reply by Leonidas-ai** [reply]
Whatever you do, do it more efficiently than what we've got now.

**Reply by Working-Joe** [reply]
**Let's start with 3:**

Of course it's logical that multiple components controlling the same input pin leads to unexpected behaviour, but it's *possible* and it may happen when you don't expect it. Preventing these errors by design adds to the intuitiveness of the system. This is not the main problem I'm aiming to solve but it's an added side-effect. Using the combine component *adds* functionality since you can then reliably combine signals where it currently can lead to unexpected behaviour as  it is dependent on update order of the components. 

Classic examples are people using two levers on `set_state` inputs, or receiving multiple signals on the same WiFi channel. 

On that, WiFi is something I still need to consider now that I come to think of it since you can't restrict this to one input per channel (or combine it with deterministic probability). Perhaps the closest signal would be prioritised, but I don't like positional dependency... That being said if the new components would live in circuit box, the WiFi component may not be need to be reworked (since all comps are in the same place) and it could be considered as an external machine instead.

**Then for 1, 4, and 5:**

The main problems that I'm aiming to solve are the inconcistencies arising from time frames, continuous vs. pulsed signals and update order dependencies. To solve this, components are unified to all output as soon as they receive both inputs. If you wouldn't use null signals, this would mean that some components do not send a (new) output. This would be similar to making all components pulsed. 

A relay for example would send it's signal as soon as it has received the (connected) signal inputs *and* the connected state controlling signal. 

However, this would become a problem if you try to combine signals. If you allow receiving multiple signals on one pin, you'd run into unexpected behaviour. Even if you'd an output as soon as you receive two signals you could still receive more signals that you would have to discard, leading to unexpected (update order dependent) behaviour again. The only deterministic solution I could come up with is to use null signals. Then all connected inputs will receive an input, it's trivial to decide when to send a signal and you limit a component to send only one signal per frame.

It is then not possible to have all components pulsed *and* have the system entirely update order independent. So it's a trade-off between robustness/intuitiveness and performance, where I think the intuitiveness of having all comps continuous (but limited) is preferable since the performance decrease may be marginal. 

It is also not possible to do this like edge-detection like fasion, since you would not no if a component will receive an input in advance, so you can't immediately decide to send or not. This means that you can't send as soon as you receive a signal.

**And for 6:**

Implementing this system without keeping all existing components would certainly break a lot, even if just because of the restriction in the amount of connections. I think if the solution is added along with circuit boxes this would definitely be optimal.

**Reply by mLuby** [reply]
Thanks for the thoughtful response.

For 3. did you mean limiting to _one wire per pin_? I thought you meant enforcing _one input pin per component_. The former I can support (and can understand how `merge`/`split` components would become useful); the latter proposal I still find confusing—maybe a picture would help, if that's indeed what you proposed.

> To solve this, components are unified to all output as soon as they receive both inputs.
Does "unifying to all output" mean updating each component which has all its inputs until there are no more differences between inputs and outputs (i.e. the network is in a stable state)?

I think what you're suggesting (always sending null unless there's a value) looks like this, where `n` = null:
```
Mem1(1).out->in1.Add0
Btn3.out->in.Sig3(in==1?3:n).out->in2.Add0
Btn5.out->in.Sig5(in==1?5:n).out->in2.Add0
```
```
ITEM   T==0   T==1   T==2   T==3
Mem1   n->1   n->1   n->1   n->1
Add0 1+n->n 1+3->4 1+5->6 1+?->?
Btn3   n->n   n->1   n->n   n->1
Sig3   n->n   1->3   n->n   1->3
Btn5   n->n   n->n   n->1   n->1
Sig5   n->n   n->n   1->5   1->5
```
If that's an accurate representation, how does your proposal handle `T==3`, when there are two non-null signals on one input pin? Also, are all the updates happening within one tick/update (as above) to avoid order dependencies?

**Reply by Working-Joe** [reply]
> did you mean limiting to one wire per pin?

Woops, that's exactly what I meant yeah.

> Does "unifying to all output" mean updating each component which has all its inputs until there are no more differences between inputs and outputs (i.e. the network is in a stable state)?

Not really, with unifying I mean that each component behaves the same - namely that they send the output on receiving both inputs (so not in the update method).
Then all source components are updated one by one.  This source then sends out its signal to the next component. This component receives the signal, and if it has already received the other inputs (or has 1 input) it will send its output. So the signal propagates until it reaces a component which still expects another input. 

The updates all happen in the same tick. It will be impossible to have two signals on the same pin since only one input connection is allowed, but if it were to occur the second signal is simply ignored.

Taking your example but removing the forbidden connection:

```
Mem1(1).out->in1.Add0
Btn3.out->in.Sig3(in==1?3:n).out->in2.Add0
```

Would update something like this:

- Mem1 is updated, sets in1.Add0 to 1
- The adder does not yet have Add1 so propagation stops

- Btn3 is updated, sets in1.Sig3 to 1.
- Sig3 has all expected inputs, outputs 3 to in2.add0
- Add0 has all inputs, outputs 4

Comps can only send one output per frame so any anomalous further received signals will not result in more outputs.

**Comment by Udrakan** [comment]
Circuit boxes being purchasable at stores would be excellent. I imagine there would be some vanilla ones, as well as circuit box mod collections. For example O2 power regulation box that will switch the generator on and off regularly (potentially causing overvoltage, but thats part of the interactivity and fun), burst-fire mode for coilgun circuit box and of course a couple reactor circuit boxes.

**Comment by PaladynMinak** [comment]
@Working-Joe 
> * Any arithmetic component for some reason (bug?) starts with their inputs true for the duration of their timeframe. This can not be patched easily because it currently provides the functionality of detecting a round start.

I haven't checked if relay resets to enabled state between game loads, but I made setup, which I've tested, that works on every game load.

![barotrauma simple level timmer](https://user-images.githubusercontent.com/58741172/133582732-bfeace7e-4270-42ac-b7ae-dd7d03516fca.png)

**Reply by Working-Joe** [reply]
Oh nice! This will work in-between rounds because signals in the delays are not serialised, so when a new round starts the delays are empty again. 

It's a bit hacky though because it seems to me that it working depends on the update order. Once the delays are over both delays are trying to control the state. For this to work the delay with the 0 must send it's output before the relay sends its output and after the delay with the 1 sends it's update. So you'd have to place the components in that order for it to work.

Still proven me wrong though :) nothing wrong with a circuit that relies on update order.

**Reply by PaladynMinak** [reply]
If I get it right every item is updated every frame, but in the order they were added? So with current standard of at least 60 fps, time frame I've set in delays, should let whole setup have enough frames to work as intended even if made from components with not setup wise id order?

**Reply by Working-Joe** [reply]
Yes, each frame every component updates in the order of their IDs. And items are assigned IDs in the order they are spawned

**Reply by Working-Joe** [reply]
(Doesn't hold for all items though, some items send their output as soon as they receive an input. One of the inconsistencies)

**Comment by SeigneurNecron** [comment]
I'm a beginner, so I wasn't fully aware of these inconsistencies, but it does look like something that needs to be fixed and this looks like a good solution to me.

I'd just add an OR component with more than 2 pins. I've been using multiple inputs on a single pin as cheap OR.
What about multiple Wifi components sending a signal on the same frequency?

I originally posted a suggestion in the Discord for a feature very similar to the Circuit Box that is suggested here and someone linked me this discussion. Here is a copy of what I posted, to further support what is being suggested there:
> Hey, I wanna suggest adding some kind of advanced wiring box.
> Not one that would be used to distribute power in the sub, but one that would allow storing complex logic circuits that check multiple things before triggering stuff.
> I like doing these, a lot, but... even a circuit controlling a simple door can take quite a lot of space.
>
> For this reason, I'm currently hiding these circuits in the walls, in a very compact form, making them:
> - unreadable in the editor;
> - uneditable and invisible in game (which removes the ability for the player to upgrade them with more functionality or sabotage stuff if you like to play with traitors);
> - invulnerable (since they could not be repaired), not really an issue imo, but could be one for other players.
>
> To improve on this, I'd like to have a craftable item, which would probably look pretty similar to the junction box (but maybe a litte smaller) and would allow the player to:
> - define custom connections (I/O) (potentially a lot of them, like 15-20);
> - build a complex circuit inside the box, with a lot of space so you can keep the circuit easily readable.
>
> This would fix the issues mentioned above and would have the following bonus:
> - Allows to easily copy paste these circuits in the editor... or even in game, if you have all the required stuff in your inventory.
> - You could even add a feature to these in the editor: allowing the player to make a prefab of that circuit, that could be edited later (if a bug is found in the circuit for ex) and changes could be applied to all instances of the prefab.

**Comment by 3e849f2e5c** [comment]
Big fan of the writeup and mockups, I might have a look at circuit boxes at some point in the near future unless the game is going to need a full rewrite to support this

**Comment by Marutzo** [comment]
I'd love this,  full support for the circuit boxes and the proposed changes, hopefully it can be done. :o)

**Comment by Swato85** [comment]
Its like every subbuilder comes to the idea of circuit boards/boxes :-D. Waiting for it apx.  2 years already. Definitely a must for this game.

**Comment by BlackMoons** [comment]
Circuit boxes that could take damage could provide new avenues for automation to fail. Maybe even be purposefully neglected for reasons. 

Also on a subnote, 'no signal' does exist in electronics (High-z output), however often it exists in conjunction with a 'pullup/down' that sets the 'default' value of a signal, that can be easily overridden by any strong signal.

**Comment by Saevon** [comment]
I like the idea of Circuit Boxes using just "one" type of wire, and "one" type of component, and just having their own UI instead of an inventory.

It would mean the UI can be customized in a logical, easy to use manner, rather then the inventory. AND it can then be used for sub-building too.

currently having to hold on to like 20 components of various types really hurts creativity, especially in the moment. 


P.S. See #4721  for my Circuit board suggestion.  I kept that part as a seperate ticket so it can have a more focused discussion. Feel free to add some more there too!

**Comment by LaysDragon** [comment]
I noticed the weird circuit reaction that made me headache ,and trying to find out how this system actually work.Just like what I guess. Hope the game will rework 
it soon,its made the circuit behavior unpredictable as a game core element.
Also love the idea of the Circuit Boxes idea!

**Comment by FalcoGer** [comment]
> Allow one wire per input pin on components.

This is tricky because I for example have a circuit that can set a reactor's temperature to any value I desire. It has 2 modes, one to set the temperature at 5950 to just get as much depleted fuel as possible, and one to set the temperature value to the precalculated, optimum, minimum value possible without loosing output. To achieve this, I can simply wire both inputs into the same pin and turn off whichever input I don't want at the time. If only one input were allowed per pin, I would have to multiply by 0 and 1 respectively and then add them together, which just makes the whole thing more complicated.

**Comment by pieslorde** [comment]
I'd like to add a few suggestions for additional components that I believe would make things a bit easier and cleaner for players, albeit maybe increasing the learning-curve for new players. One example is the multiply block, which I find I am often feeding one of the inputs with a memory block. This makes a real circuit, an amplifier, take twice the materials and space as a component like the Factorial Component. I think this isn't ideal both because the amplifier is a circuit which only takes a handful of transistors (10~100) and two resistors, whereas performing a factorial isn't a practical analog circuit at all and would require quite a bit of area on an FPGA or ASIC, and finding the gain of an amplifier with our current wiring system requires that the builder find the multiplier, figure out which memory component is setting the gain, and then change the gain there.

In addition to the case of a fixed-gain amplifier, I think including blocks which represent common analog circuits used in the real world would make building traditional linear circuits (like the ones used in ARCs and ABCs) less complicated and easier for people to learn because it would be the exact same symbols and theory used in educational material outside of Barotrauma. These are my suggested additions:

- Gain block (outputs a fixed, configurable multiple of the input)
- Bias block (outputs the input added with a fixed, configurable value)
- Differentiator block (avoids the messy-ness of using delay blocks, which don't output the true slope of signals since they are comparing samples from potentially non-adjacent frames)
- Integrator block (again avoiding the high-complexity of what is, in real life, an operational amplifier with a resistor and capacitor)
- LPF block (just like the IIR filters in Reactor.cs that consist of a single MathUtils.Lerp function, except now the user can choose a fixed cutoff frequency to smooth out a signal)
- HPF block (same IIR filter as the LPF, but the IIR coefficient is negative)

The wiring system as it exists is presented to the player as a continuous-time system. The current solution to filtering, differentiating, and integrating is to build them as discrete-time systems within a pseudo-continuous-time system that is actually a discrete-time system. Adding the 4 last components in the list would allow the player to build any linear time-invariant system using a small number of components without having to worry about sample times. The actual discrete-filter calculation would be handled in the background and presented to the player as if it is a traditional continuous-time analog circuit.

An additional component that would solve FalcoGer's very astute criticism of the "one wire per input" suggestion would be to include multiplexers. Many of the current components have inherent multiplexers (the And, Equals, Greater, Or, RegEx Find, Signal Check, and Xor Components) which multiplex between a fixed value and an optional input, but we do not have a block which can select between two non-fixed inputs. A multiplexer would basically be a Signal Check Component with a "SET_FALSE_OUTPUT" input.

I would also like to mention that the Equals and Signal Check components have functional overlap, or at least I believe it to be the case. The SET_TARGETSIGNAL input of the Signal Check componnet is the same as the SIGNAL_IN_2 input of the Equals component. Perhaps it would be beneficial to unify the comparison components (Greater, Equals, Signal Check) into a single component much like the Oscillator component, such that the user can select between behaviour with a drop-down menu. Similarly, the trig components could be condensed into a configurable component.

**Reply by mygamingaccount** [reply]
Gain block is a multiply component that stores one of its inputs

A much smaller change than the scope of this topic

**Reply by mygamingaccount** [reply]
Filter blocks can be made with components if you are careful about the order of the items

Video of a 0.1Hz Butterworth LPF
https://cdn.discordapp.com/attachments/903478852259573760/1031618668557246464/2022-10-17_11-37-24.mp4

**Reply by mygamingaccount** [reply]
I'm curious, what use would you have for a HPF in a submarine?

**Reply by pieslorde** [reply]
I like that you made a separate discussion thread for the constant-input case for blocks. I think that would be a better solution than adding new blocks with more overlapping functionality.

I know that filters can be made in game at the moment, but right now it is very costly to do so and as far as I am aware, the player has to be at least a little familiar with digital signal processing fundamentals to do it (sampling theorem, conversion from the Z-domain to S-domain, filter stability critereon, etc.). I guess it's possible to not take any of that into consideration, but I think eye-balling the filter response is more trouble than it's worth when we could add a block that can be configured using real-world units.

As for the use-case of a HPF, the Reactor has a 5%/sec slew limiter in series with a 1/(2*pi) Hz 1st-order LPF on the fission and turbine inputs. The optimal inputs to this system, ignoring the non-linear slew limiter element, would be the DC response passed through a differentiator summed with the DC response. To avoid giving too much control to the high-frequency components, a HPF multiplied by some gain and summed with the ideal DC input would counteract most of the intrinsic LPF in the reactor.

Additionally, the HPF would be useful for battery controllers since it could be used to determine the transient power network deficit, thus helping determine the ideal charging rate. Personally though, I agree with your point in #6870 that the way batteries work in this game are fundamentally different from how they work IRL. I would love to see functionality added to the relay component (or a new component) that operates like a transistor or current-limiting device. This would make it possible to gradually brown-out subcircuits on the sub that aren't as important, and of course it would allow the battery charge and discharge current to be regulated by an external circuit while allowing the batteries to more closely match IRL batteries. Without these circuits, over-current in the battery would simply damage it, potentially causing fires and/or explosions.

I might be alone on this opinion, but the more the wiring system can be analogized to real-life stuff, the easier it might be for players to understand what the components are doing.

Idk if this is the wrong place to be writing all of this, so if any or all of this is out of scope here, I can start a new discussion thread.

**Comment by mygamingaccount** [comment]
These circuit boxes might be the devices themselves:

Hook up components _inside_ the door's connection panel, or the nav terminal, or the engine, etc.
It is already a hassle to hide wiring components. They are usually hidden inside walls. If you could store them in the wiring panels it would make a lot of sense.

**Reply by pieslorde** [reply]
This is a great idea, however in Working-Joe's initial writeup they explained that the enclosures would only hold a certain number of components so that more complex systems are more economically costly than simple ones. Perhaps the wiring panels of non-components could have a slot to accept a circuit enclosure/board?

Are you suggesting to only have circuit boxes inside the wiring panels of machines, or to have that functionality in addition to having circuit boxes that can be placed on walls like a regular wiring component?

**Comment by DarthPointer** [comment]
I'll drop my ideas here in case anyone cares.

Manifesto:

0. Keep all the existing signal items as they are. Make them update strictly either before or after the new signals. With all the quirks, so that the legacy structures work and that there is an alternative in case the new system fucks up.
1. Maintain the joined inputs: plugging multiple signals into one input will make the components use a non-null signal (if any, null otherwise).
2. Maintain the single-tick calculations: inputs to the element inputs take effect on its outputs once and this same tick they were sent. Exception: memory component, which should have one tick overwrite delay as a feature (and mention that darn explicitly).
3. Make the circuits behave invariant to installation order.
4. Standardise the element input defaults and resetting configuration options.
5. Add more out pins to monitor boat devices. Things like actual power consumption, for example. There is no use to mirroring charge_rate_in of a battery, while having data on currently consumed KW, max consumed/produced KW (including boat upgrade stats) would be really useful.
6. Add smooth control toggles. Yes, I dislike batteries only having 6 charge rate position sliders. The floating point numbers exist. I don't want to wire enormous webs to assure smooth battery recharge KW changes.
7. Add ingame-editable tags for all the copmonents and circuit box pins. Display the custom name tag (if any) as the item/box component name with its type name as a comment (**Total Power** adder component).
8. Add wire pins and join points in the circuit boxes. The wire mess caused by lack of them is too much of Quality of Unlife to keep it for arbitrary limitations like 5 wires per component pin.
9. Add constant value configuration UI for things like buttons and sensors to configure true/false/pulse/ambient values of these.
10. A toggle to persist state of mem and delay elements between scene loads. This will both let players free of the routine of restoring circuitry state if needed, and effectively provide a scene start pulse solution.


Implementation:
1. Each signal sent by a component is either an error, or a value. Both can have metadata attached, same way as signals can have extra data passed for uses like turret operator tracking.
2. An error consists of an error string and original element id. Errors propagate through elements, persisting the error string and source id.
3. Dependency loops are not allowed. If it is detected that an "instant effect" output (basically all inputs but mem value and delay component value when not set to 0 delay) of an element affects on its input, the element sets its outputs to a "Dependency Loop" error.
If an element has already been set to an error this frame, the first error of the frame is preserved and the new error is not propagated. Sending "null" (no signal) does NOT count as being a dependency.
4. Replace the SigCheck component with a new Multiplexer (Select, Switch, whatever works better for people with no technical background) component (bool switch, value a, value b -> value selected). Its inactive input does not propagate loop checks.
5. Standardise the configuration of all inputs. All pin inputs should have ingame configurable defaults to use if there is no non-null wire connected, with few exceptions being mem and delay val inputs.
6. For each sample/constant signal UI input add a null "no signal" checkbox. This is required if the signals are allowed to send and recieve both "nulls" and empty strings. When enabled, it fades and disables the textbox for clarity.
7. A checkbox "remember last signal" for each input pin, except for mem and delay value. When the component does not recieve a signal with this pin, it calculates its state based on last non-null input. If an error was sent and then lifted, the last non-null and non-error signal is used.
8. When "remember last signal" is set for a pin, the "no signal" checkbox of its "target signal" is disabled and reset to false.
9. [Optional] Introduce "target signal" input for each pin that is used for binary true/false state.
10. Circuit boxes process propagate dependency checks and evaluations of their inside components individualy, based on the interface pins they are connected to. Components placed into circuit boxes should behave exacactly the same
if they were placed directly and connected to outer circuitry in the same way.
11. Add 0 to null and null to 0 dedicated components. The existing ones can be used to do that, but they require configuration to do that first.
12. Copy-pasting within CB.
13. Copy-pasting across boat and CB.

10. Algorithm:
- Each frame all the components start as invalidated/not evaluated. Can be achieved by updating last frame counter each time it is updated.
- Each component updates in a fixed gameplay dt frame.
- If the component already is updated, it does nothing more.
- If not, the component runs updates depth-first through its dependencies to set its state for the current frame.

Create the set of "passed" components to detect dependency loops, add self to it.

Iterate through input pins. Components whose pins determine dependency routing start iterating with these pins: Delay components always start with delay duration pin, Switch and Relay components start with select and state pins.
.If the pin is inactive due to other pin state (switch, relay), `continue`.
.If the pin is not wired, use the set constant and `continue`.

.Iterate through wires connected to the pin.
..If the wire is not connected correctly, `continue`.

..If it is, request the value from the output pin.
...If the value `is not null`, save it and `break` the wires loop and proceed iterating through inputs.

.If the wires loop is finished with no value recieved, save null/persisted value (if persisting enabled)/default for this input.

Evaluate the outputs, internal state and mark as evaluated.  

///
Requesting an output value basically is recursion. It fires the same procedure for the dependent component, but it starts differently.

If the requested value is memory value, just return it.
If the component already is evaluated for this frame, return the relevant output.
Check if the "passed" components contain this component.
.If it is so, raise dependency loop for outputs, mark self as evaluated and return.

Run the same inputs loop.

Evaluate the outputs, internal state and mark as evaluated.
Remove self from "passed" components set.
Return the requested value.
 

Notes:
- Delay component is presumed to update its value immediately within frame if it is set to 0 delay. If this is deemed unnecessary/too complex, related specifications should be altered.
- For better flexibility, outputs of the components could be treated individually instead, only depending on relevant inputs and only propagating errors from relevant components.
- The algorithm as is assumes components inside the CBs be treated individually.
- With the existing feature of persisting last non-null input over current null input, I don't really think it is possible use random-order "voluntary" entity updates to evaluate without the enormous overheads of tracking whom waits for whom and finalization checks.
- Wifi probably should live in a separate "realm" with at least a frame of expected delay and not be used for frame-perfect applications.
- Regarding the connections check order, there still is place for "Undefined Behaviour" when multiple inputs provide non-null, but I think it can stay so. After all, it is a boon to be able to merge multiple pulse sources and process them as a single pulse source where relevant.
- Keeping all legacy signal updates strictly before/after the updates for new signals avoids UBs and complications with mixed use. Cross-connections should provide last evaluated values, no dependency shenanigans, everyone happy.
